{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# individPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Biobase)\n",
    "library(usethis)\n",
    "library(GSVAdata)\n",
    "library(hu6800.db)\n",
    "\n",
    "#### Expression data with filtered entezID\n",
    "##### get set_entrezID with matrix data\n",
    "convert_probe_to_entrezID = function(matrix_file, bioc_package){\n",
    "  #### matrix data\n",
    "  matrixdata = read.csv(matrix_file, header = TRUE, row.names = 1)\n",
    "  matrixdata = as.data.frame(matrixdata)\n",
    "  matrixdata$probe_id = rownames(matrixdata)\n",
    "  if (bioc_package == \"hu6800.db\") { probe_ENTREZID = toTable(hu6800ENTREZID)}\n",
    "  newmatrixdata = merge(probe_ENTREZID, matrixdata, by = \"probe_id\")\n",
    "  newmatrixdata$probe_id = NULL\n",
    "  newmatrixdata = aggregate(newmatrixdata[,-1], list(newmatrixdata$gene_id), FUN = median)\n",
    "  rownames(newmatrixdata) = newmatrixdata$Group.1\n",
    "  newmatrixdata = newmatrixdata[,-1]\n",
    "  newmatrixdata = as.matrix(newmatrixdata)\n",
    "  newmatrixdata\n",
    "}\n",
    "##get matrix data with filtered entrezID\n",
    "GSE1122_setentrez = convert_probe_to_entrezID(\"F:/lab_data/0. data/datasets/csv data/GSE1122_series_matrix.csv\", \"hu6800.db\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### file data as example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"F:/lab_related_R/individPath\")\n",
    "#tumorFile = \"GSE1122_disease.txt\"\n",
    "#refFile = \"GSE1122_control.txt\"\n",
    "PathwayFile = \"file.txt\"\n",
    "ControlData <- as.matrix(GSE1122_setentrez[,1:5])\n",
    "CaseData <- as.matrix(GSE1122_setentrez[,6:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load R function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> source(\"D:/individPath_1.0/individPath/R/individPath.r\") <br>\n",
    "> source(\"D:/individPath_1.0/individPath/R/individPathCal.r\")  <br>\n",
    "> source(\"D:/individPath_1.0/individPath/R/SRGgenePair.r\") <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Identifying stable and reversal intra-pathway gene pairs\"\n",
      "[1] \"Processing 1/186 : KEGG_GLYCOLYSIS_GLUCONEOGENESIS\"\n",
      "[1] \"Processing 2/186 : KEGG_CITRATE_CYCLE_TCA_CYCLE\"\n",
      "[1] \"Processing 3/186 : KEGG_PENTOSE_PHOSPHATE_PATHWAY\"\n",
      "[1] \"Processing 4/186 : KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\"\n",
      "[1] \"Processing 5/186 : KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\"\n",
      "[1] \"Processing 6/186 : KEGG_GALACTOSE_METABOLISM\"\n",
      "[1] \"Processing 7/186 : KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\"\n",
      "[1] \"Processing 8/186 : KEGG_FATTY_ACID_METABOLISM\"\n",
      "[1] \"Processing 9/186 : KEGG_STEROID_BIOSYNTHESIS\"\n",
      "[1] \"Processing 10/186 : KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\"\n",
      "[1] \"Processing 11/186 : KEGG_STEROID_HORMONE_BIOSYNTHESIS\"\n",
      "[1] \"Processing 12/186 : KEGG_OXIDATIVE_PHOSPHORYLATION\"\n",
      "[1] \"Processing 13/186 : KEGG_PURINE_METABOLISM\"\n",
      "[1] \"Processing 14/186 : KEGG_PYRIMIDINE_METABOLISM\"\n",
      "[1] \"Processing 15/186 : KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\"\n",
      "[1] \"Processing 16/186 : KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\"\n",
      "[1] \"Processing 17/186 : KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\"\n",
      "[1] \"Processing 18/186 : KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\"\n",
      "[1] \"Processing 19/186 : KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS\"\n",
      "[1] \"Processing 20/186 : KEGG_LYSINE_DEGRADATION\"\n",
      "[1] \"Processing 21/186 : KEGG_ARGININE_AND_PROLINE_METABOLISM\"\n",
      "[1] \"Processing 22/186 : KEGG_HISTIDINE_METABOLISM\"\n",
      "[1] \"Processing 23/186 : KEGG_TYROSINE_METABOLISM\"\n",
      "[1] \"Processing 24/186 : KEGG_PHENYLALANINE_METABOLISM\"\n",
      "[1] \"Processing 25/186 : KEGG_TRYPTOPHAN_METABOLISM\"\n",
      "[1] \"Processing 26/186 : KEGG_BETA_ALANINE_METABOLISM\"\n",
      "[1] \"Processing 27/186 : KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM\"\n",
      "[1] \"Processing 28/186 : KEGG_SELENOAMINO_ACID_METABOLISM\"\n",
      "[1] \"Processing 29/186 : KEGG_GLUTATHIONE_METABOLISM\"\n",
      "[1] \"Processing 30/186 : KEGG_STARCH_AND_SUCROSE_METABOLISM\"\n",
      "[1] \"Processing 31/186 : KEGG_N_GLYCAN_BIOSYNTHESIS\"\n",
      "[1] \"Processing 32/186 : KEGG_OTHER_GLYCAN_DEGRADATION\"\n",
      "[1] \"Processing 33/186 : KEGG_O_GLYCAN_BIOSYNTHESIS\"\n",
      "[1] \"Processing 34/186 : KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\"\n",
      "[1] \"Processing 35/186 : KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\"\n",
      "[1] \"Processing 36/186 : KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\"\n",
      "[1] \"Processing 37/186 : KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\"\n",
      "[1] \"Processing 38/186 : KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\"\n",
      "[1] \"Processing 39/186 : KEGG_GLYCEROLIPID_METABOLISM\"\n",
      "[1] \"Processing 40/186 : KEGG_INOSITOL_PHOSPHATE_METABOLISM\"\n",
      "[1] \"Processing 41/186 : KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\"\n",
      "[1] \"Processing 42/186 : KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\"\n",
      "[1] \"Processing 43/186 : KEGG_ETHER_LIPID_METABOLISM\"\n",
      "[1] \"Processing 44/186 : KEGG_ARACHIDONIC_ACID_METABOLISM\"\n",
      "[1] \"Processing 45/186 : KEGG_LINOLEIC_ACID_METABOLISM\"\n",
      "[1] \"Processing 46/186 : KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\"\n",
      "[1] \"Processing 47/186 : KEGG_SPHINGOLIPID_METABOLISM\"\n",
      "[1] \"Processing 48/186 : KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\"\n",
      "[1] \"Processing 49/186 : KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES\"\n",
      "[1] \"Processing 50/186 : KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\"\n",
      "[1] \"Processing 51/186 : KEGG_PYRUVATE_METABOLISM\"\n",
      "[1] \"Processing 52/186 : KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\"\n",
      "[1] \"Processing 53/186 : KEGG_PROPANOATE_METABOLISM\"\n",
      "[1] \"Processing 54/186 : KEGG_BUTANOATE_METABOLISM\"\n",
      "[1] \"Processing 55/186 : KEGG_ONE_CARBON_POOL_BY_FOLATE\"\n",
      "[1] \"Processing 56/186 : KEGG_RIBOFLAVIN_METABOLISM\"\n",
      "[1] \"Processing 57/186 : KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\"\n",
      "[1] \"Processing 58/186 : KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\"\n",
      "[1] \"Processing 59/186 : KEGG_FOLATE_BIOSYNTHESIS\"\n",
      "[1] \"Processing 60/186 : KEGG_RETINOL_METABOLISM\"\n",
      "[1] \"Processing 61/186 : KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\"\n",
      "[1] \"Processing 62/186 : KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\"\n",
      "[1] \"Processing 63/186 : KEGG_LIMONENE_AND_PINENE_DEGRADATION\"\n",
      "[1] \"Processing 64/186 : KEGG_NITROGEN_METABOLISM\"\n",
      "[1] \"Processing 65/186 : KEGG_SULFUR_METABOLISM\"\n",
      "[1] \"Processing 66/186 : KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\"\n",
      "[1] \"Processing 67/186 : KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\"\n",
      "[1] \"Processing 68/186 : KEGG_DRUG_METABOLISM_CYTOCHROME_P450\"\n",
      "[1] \"Processing 69/186 : KEGG_DRUG_METABOLISM_OTHER_ENZYMES\"\n",
      "[1] \"Processing 70/186 : KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\"\n",
      "[1] \"Processing 71/186 : KEGG_ABC_TRANSPORTERS\"\n",
      "[1] \"Processing 72/186 : KEGG_RIBOSOME\"\n",
      "[1] \"Processing 73/186 : KEGG_RNA_DEGRADATION\"\n",
      "[1] \"Processing 74/186 : KEGG_RNA_POLYMERASE\"\n",
      "[1] \"Processing 75/186 : KEGG_BASAL_TRANSCRIPTION_FACTORS\"\n",
      "[1] \"Processing 76/186 : KEGG_DNA_REPLICATION\"\n",
      "[1] \"Processing 77/186 : KEGG_SPLICEOSOME\"\n",
      "[1] \"Processing 78/186 : KEGG_PROTEASOME\"\n",
      "[1] \"Processing 79/186 : KEGG_PROTEIN_EXPORT\"\n",
      "[1] \"Processing 80/186 : KEGG_PPAR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 81/186 : KEGG_BASE_EXCISION_REPAIR\"\n",
      "[1] \"Processing 82/186 : KEGG_NUCLEOTIDE_EXCISION_REPAIR\"\n",
      "[1] \"Processing 83/186 : KEGG_MISMATCH_REPAIR\"\n",
      "[1] \"Processing 84/186 : KEGG_HOMOLOGOUS_RECOMBINATION\"\n",
      "[1] \"Processing 85/186 : KEGG_NON_HOMOLOGOUS_END_JOINING\"\n",
      "[1] \"Processing 86/186 : KEGG_MAPK_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 87/186 : KEGG_ERBB_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 88/186 : KEGG_CALCIUM_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 89/186 : KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\"\n",
      "[1] \"Processing 90/186 : KEGG_CHEMOKINE_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 91/186 : KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\"\n",
      "[1] \"Processing 92/186 : KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\"\n",
      "[1] \"Processing 93/186 : KEGG_CELL_CYCLE\"\n",
      "[1] \"Processing 94/186 : KEGG_OOCYTE_MEIOSIS\"\n",
      "[1] \"Processing 95/186 : KEGG_P53_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 96/186 : KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\"\n",
      "[1] \"Processing 97/186 : KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\"\n",
      "[1] \"Processing 98/186 : KEGG_REGULATION_OF_AUTOPHAGY\"\n",
      "[1] \"Processing 99/186 : KEGG_LYSOSOME\"\n",
      "[1] \"Processing 100/186 : KEGG_ENDOCYTOSIS\"\n",
      "[1] \"Processing 101/186 : KEGG_PEROXISOME\"\n",
      "[1] \"Processing 102/186 : KEGG_MTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 103/186 : KEGG_APOPTOSIS\"\n",
      "[1] \"Processing 104/186 : KEGG_CARDIAC_MUSCLE_CONTRACTION\"\n",
      "[1] \"Processing 105/186 : KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\"\n",
      "[1] \"Processing 106/186 : KEGG_WNT_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 107/186 : KEGG_DORSO_VENTRAL_AXIS_FORMATION\"\n",
      "[1] \"Processing 108/186 : KEGG_NOTCH_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 109/186 : KEGG_HEDGEHOG_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 110/186 : KEGG_TGF_BETA_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 111/186 : KEGG_AXON_GUIDANCE\"\n",
      "[1] \"Processing 112/186 : KEGG_VEGF_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 113/186 : KEGG_FOCAL_ADHESION\"\n",
      "[1] \"Processing 114/186 : KEGG_ECM_RECEPTOR_INTERACTION\"\n",
      "[1] \"Processing 115/186 : KEGG_CELL_ADHESION_MOLECULES_CAMS\"\n",
      "[1] \"Processing 116/186 : KEGG_ADHERENS_JUNCTION\"\n",
      "[1] \"Processing 117/186 : KEGG_TIGHT_JUNCTION\"\n",
      "[1] \"Processing 118/186 : KEGG_GAP_JUNCTION\"\n",
      "[1] \"Processing 119/186 : KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\"\n",
      "[1] \"Processing 120/186 : KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\"\n",
      "[1] \"Processing 121/186 : KEGG_RENIN_ANGIOTENSIN_SYSTEM\"\n",
      "[1] \"Processing 122/186 : KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 123/186 : KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 124/186 : KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 125/186 : KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\"\n",
      "[1] \"Processing 126/186 : KEGG_JAK_STAT_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 127/186 : KEGG_HEMATOPOIETIC_CELL_LINEAGE\"\n",
      "[1] \"Processing 128/186 : KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\"\n",
      "[1] \"Processing 129/186 : KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 130/186 : KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 131/186 : KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 132/186 : KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\"\n",
      "[1] \"Processing 133/186 : KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\"\n",
      "[1] \"Processing 134/186 : KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\"\n",
      "[1] \"Processing 135/186 : KEGG_CIRCADIAN_RHYTHM_MAMMAL\"\n",
      "[1] \"Processing 136/186 : KEGG_LONG_TERM_POTENTIATION\"\n",
      "[1] \"Processing 137/186 : KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 138/186 : KEGG_LONG_TERM_DEPRESSION\"\n",
      "[1] \"Processing 139/186 : KEGG_OLFACTORY_TRANSDUCTION\"\n",
      "[1] \"Processing 140/186 : KEGG_TASTE_TRANSDUCTION\"\n",
      "[1] \"Processing 141/186 : KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\"\n",
      "[1] \"Processing 142/186 : KEGG_INSULIN_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 143/186 : KEGG_GNRH_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 144/186 : KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\"\n",
      "[1] \"Processing 145/186 : KEGG_MELANOGENESIS\"\n",
      "[1] \"Processing 146/186 : KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\"\n",
      "[1] \"Processing 147/186 : KEGG_TYPE_II_DIABETES_MELLITUS\"\n",
      "[1] \"Processing 148/186 : KEGG_TYPE_I_DIABETES_MELLITUS\"\n",
      "[1] \"Processing 149/186 : KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\"\n",
      "[1] \"Processing 150/186 : KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\"\n",
      "[1] \"Processing 151/186 : KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\"\n",
      "[1] \"Processing 152/186 : KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\"\n",
      "[1] \"Processing 153/186 : KEGG_ALZHEIMERS_DISEASE\"\n",
      "[1] \"Processing 154/186 : KEGG_PARKINSONS_DISEASE\"\n",
      "[1] \"Processing 155/186 : KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\"\n",
      "[1] \"Processing 156/186 : KEGG_HUNTINGTONS_DISEASE\"\n",
      "[1] \"Processing 157/186 : KEGG_PRION_DISEASES\"\n",
      "[1] \"Processing 158/186 : KEGG_VIBRIO_CHOLERAE_INFECTION\"\n",
      "[1] \"Processing 159/186 : KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\"\n",
      "[1] \"Processing 160/186 : KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\"\n",
      "[1] \"Processing 161/186 : KEGG_LEISHMANIA_INFECTION\"\n",
      "[1] \"Processing 162/186 : KEGG_PATHWAYS_IN_CANCER\"\n",
      "[1] \"Processing 163/186 : KEGG_COLORECTAL_CANCER\"\n",
      "[1] \"Processing 164/186 : KEGG_RENAL_CELL_CARCINOMA\"\n",
      "[1] \"Processing 165/186 : KEGG_PANCREATIC_CANCER\"\n",
      "[1] \"Processing 166/186 : KEGG_ENDOMETRIAL_CANCER\"\n",
      "[1] \"Processing 167/186 : KEGG_GLIOMA\"\n",
      "[1] \"Processing 168/186 : KEGG_PROSTATE_CANCER\"\n",
      "[1] \"Processing 169/186 : KEGG_THYROID_CANCER\"\n",
      "[1] \"Processing 170/186 : KEGG_BASAL_CELL_CARCINOMA\"\n",
      "[1] \"Processing 171/186 : KEGG_MELANOMA\"\n",
      "[1] \"Processing 172/186 : KEGG_BLADDER_CANCER\"\n",
      "[1] \"Processing 173/186 : KEGG_CHRONIC_MYELOID_LEUKEMIA\"\n",
      "[1] \"Processing 174/186 : KEGG_ACUTE_MYELOID_LEUKEMIA\"\n",
      "[1] \"Processing 175/186 : KEGG_SMALL_CELL_LUNG_CANCER\"\n",
      "[1] \"Processing 176/186 : KEGG_NON_SMALL_CELL_LUNG_CANCER\"\n",
      "[1] \"Processing 177/186 : KEGG_ASTHMA\"\n",
      "[1] \"Processing 178/186 : KEGG_AUTOIMMUNE_THYROID_DISEASE\"\n",
      "[1] \"Processing 179/186 : KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\"\n",
      "[1] \"Processing 180/186 : KEGG_ALLOGRAFT_REJECTION\"\n",
      "[1] \"Processing 181/186 : KEGG_GRAFT_VERSUS_HOST_DISEASE\"\n",
      "[1] \"Processing 182/186 : KEGG_PRIMARY_IMMUNODEFICIENCY\"\n",
      "[1] \"Processing 183/186 : KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\"\n",
      "[1] \"Processing 184/186 : KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\"\n",
      "[1] \"Processing 185/186 : KEGG_DILATED_CARDIOMYOPATHY\"\n",
      "[1] \"Processing 186/186 : KEGG_VIRAL_MYOCARDITIS\"\n",
      "[1] \"individPath_processing 1/4 : GSM18409\"\n",
      "[1] \"individPath_processing 2/4 : GSM18410\"\n",
      "[1] \"individPath_processing 3/4 : GSM18411\"\n",
      "[1] \"individPath_processing 4/4 : GSM18412\"\n"
     ]
    }
   ],
   "source": [
    "SampleInfo <- colnames(CaseData)[-1]\n",
    "NumSample <- length(SampleInfo)\n",
    "PathData <- read.table(PathwayFile, header=F,sep=\"\\t\",fill=T)\n",
    "PathName <- as.matrix(PathData[,1])\n",
    "OutPath <- paste(\"individPath_Result_\", Sys.Date(),sep=\"\")\n",
    "dir.create(OutPath)\n",
    "\n",
    "###---------Identifying Stable gene pair -------###\n",
    "print(\"Identifying stable and reversal intra-pathway gene pairs\")\n",
    "cutoff <- 0.99\n",
    "GP.result <- SRGgenePair(ControlData, CaseData, PathData, cutoff)\n",
    "StableGP <- GP.result$BG.GenePairs\n",
    "NumStable <- nrow(StableGP)\n",
    "ReversalGP <- GP.result$ReversalStat\n",
    "PathGP <- GP.result$PathGP\n",
    "\n",
    "###---------Individualized altered pathway -------###\n",
    "Result <- NULL\n",
    "AD.result <- NULL\n",
    "for( i in 1:NumSample){\n",
    "  print(paste(\"individPath_processing \", i, \"/\", NumSample,\" : \", SampleInfo[i], sep=\"\"))\n",
    "  patient <- CaseData[,i+1]\n",
    "  names(patient) <- CaseData[,1]\n",
    "  \n",
    "  Result.tmp <- individPathCal(patient, StableGP, ReversalGP, NumStable, PathGP)\n",
    "  Result <- cbind(Result,Result.tmp)\n",
    "  AD.p <- as.matrix(p.adjust(Result.tmp, \"BH\"))\n",
    "  AD.result <- cbind(AD.result,AD.p)\n",
    "}\n",
    "Last.Result <- cbind( PathName, Result)\n",
    "Last.ADResult <- cbind( PathName, AD.result)\n",
    "\n",
    "colnames(Last.Result)[2:(NumSample+1)] <- SampleInfo\n",
    "colnames(Last.Result)[1] <- \"PathwayID\"\n",
    "colnames(Last.ADResult)[2:(NumSample+1)] <- SampleInfo\n",
    "colnames(Last.ADResult)[1] <- \"PathwayID\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "###export results\n",
    "write.table(Last.Result, file=paste(OutPath,\"/individPath_BH_result.txt\",sep=\"\"),\n",
    "            row.names=FALSE, col.names=F, sep=\"\\t\",quote=FALSE)\n",
    "write.table(Last.ADResult, file=paste(OutPath,\"/individPath_pvalue_result.txt\",sep=\"\"),\n",
    "            row.names=FALSE, col.names=F, sep=\"\\t\",quote=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.5.3 (2019-03-11)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 17763)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=English_United States.1252 \n",
       "[2] LC_CTYPE=English_United States.1252   \n",
       "[3] LC_MONETARY=English_United States.1252\n",
       "[4] LC_NUMERIC=C                          \n",
       "[5] LC_TIME=English_United States.1252    \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    parallel  stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] individPath_1.0      hu6800.db_3.2.3      GSVAdata_1.18.0     \n",
       " [4] hgu95a.db_3.2.3      org.Hs.eg.db_3.6.0   GSEABase_1.42.0     \n",
       " [7] graph_1.58.2         annotate_1.58.0      XML_3.98-1.19       \n",
       "[10] AnnotationDbi_1.42.1 IRanges_2.14.12      S4Vectors_0.18.3    \n",
       "[13] usethis_1.5.0        Biobase_2.40.0       BiocGenerics_0.26.0 \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] Rcpp_1.0.1      pillar_1.3.1    compiler_3.5.3  base64enc_0.1-3\n",
       " [5] bitops_1.0-6    tools_3.5.3     digest_0.6.18   uuid_0.1-2     \n",
       " [9] bit_1.1-14      jsonlite_1.6    RSQLite_2.1.1   evaluate_0.13  \n",
       "[13] memoise_1.1.0   pkgconfig_2.0.2 rlang_0.3.4     IRdisplay_0.7.0\n",
       "[17] DBI_1.0.0       IRkernel_1.0.1  repr_1.0.0      fs_1.3.1       \n",
       "[21] bit64_0.9-7     glue_1.3.1      pbdZMQ_0.3-3    blob_1.1.1     \n",
       "[25] htmltools_0.3.6 xtable_1.8-4    RCurl_1.95-4.12 crayon_1.3.4   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.5.3",
   "language": "R",
   "name": "ir353"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
