# A collection of R tools for benchmarking genomic-range GSA methods
First version: May 25th, 2020<br>
Last review: May 27th, 2020
## What is genomic-range GSA?
A genomic-range GSA is a process of seeking functional enrichment for the genomic regions. This goes a tier deeper than mapping genes to pathways, and we now map genomic regions to genes and then to pathways. Typically the results from **Gene Ontology (GO)** and **Kyoto Encyclopedia of Genes and Genomes (KEGG)** are sought for the same.<br>
## Contents
The **R** folder contains the functions in-use for the example pipeline.<br>
The **data** folder contains the intermediary files that are saved during the execution cycle of the pipeline.<br>
The **man** folder contains the definitions and usage for the functions in-use.<br>
The **results** folder contains resultant tables and plots generated from the execution.<br>
The **workflows** folder contains the jupyter notebook for the example pipepine and relevant data.<br>
## References
See our papers:<br><br>
XIE, C., JAUHARI, S., MORA, A. (2020), Popularity and performance of bioinformatics software -The case of gene set analysis <br>
MORA, A. (2019), Gene set analysis methods for the functional interpretation of non-mRNA dataâ€”Genomic range and ncRNA data, Briefings in Bionformatics, doi:10.1093/bib/bbz090
